src/atoms/affine/multiply_divide.jl

```diff
@@ -176,6 +203,42 @@ function conic_form!(x::DotMultiplyAtom, unique_conic_forms::UniqueConicForms)
     return get_conic_form(unique_conic_forms, x)
 end
 
+function template(x::DotMultiplyAtom, context)
+    subproblems = template.(children(x), Ref(context))
+
+    @assert length(children(x)) == length(subproblems) == 2
+    s1, s2 = subproblems
+    if vexity(x.children[1]) != ConstVexity()
+        if vexity(x.children[2]) != ConstVexity()
+            error("multiplication of two non-constant expressions is not DCP compliant")
+        else
+            # make sure first child is the one that's constant
+            x.children[1], x.children[2] = x.children[2], x.children[1]
+            s1, s2 = s2, s1
+        end
+    end
+
+    # promote the size of the coefficient matrix, so eg
+    # 3 .* x
+    # works regardless of the size of x
+    coeff = evaluate(x.children[1]) .* ones(size(x.children[2]))
+    # promote the size of the variable
+    # we've previously ensured neither x nor y is 1x1
+    # and that the sizes are compatible,
+    # so if the sizes aren't equal the smaller one is size 1
+    var = x.children[2]
+    if size(var, 1) < size(coeff, 1)
+        var = ones(size(coeff, 1)) * var
+    elseif size(var, 2) < size(coeff, 2)
+        var = var * ones(1, size(coeff, 1))
+    end
+
+    const_multiplier = Diagonal(vec(coeff))
```

Comment:
I copied this method from the `conic_form!` one (but using `Diagonal` instead of `spdiagm`, which I thought might be more efficient here?). Convex.jl uses a lot of these "vectorized" methods where we create a big matrix to act on the vector of variables. I wonder if there are better ways that don't involve constructing giant matrices.